ADA/ada.qmd

```diff
@@ -19,57 +17,248 @@ It brings together:
+
+---
+
+#### Writing and running code: workflows
+
+---
+
+Workflows are a Databricks feature that allow you to schedule and automatically run your code at specified times and in the order you suggest. 
+
+Workflows allow you to automate the running of single or multiple scripts, so if you have large amounts of data processing to do on a regular basis, you can set up a workflow to complete this overnight. Your laptop does not need to be switched on for this to work. 
+
+
+---
+
+#### Writing and running code: notebooks
+
+---
+
+Notebooks might be something you've not come across before. Notebooks are made up of multiple code or markdown cells, which may mean that you have to structure your code in a different way than you're used to. They are different to IDEs (Integrated Development Environments) like RStudio. You can find more information on Databricks notebooks and their structure on the [ADA guidance site](https://analytical-data-access.education.gov.uk/get-support/using-r-python-and-sql-in-databricks).
+
+Notebooks can be helpful for ad hoc, exploratory, or one off pieces of analysis to add additional documentation as you go. You can also use notebooks to present findings to others by hiding code cells and only showing Markdown cells containing charts, tables, images or text. They can be exported as HTML files to enable easier sharing outside of the Databricks interface. 
+
+:::{.callout-important}
+If the rest of your team are unfamiliar with notebooks and will need to pick up your work at a later date, use of notebooks without proper handover or documentation could cause issues and result in work no longer being reproducible. The [Statistics Development Team](mailto:statistics.development@education.gov.uk) are happy to advise on this and provide guidance if you have any concerns. 
+:::
+
+Due to reasons highlighted above, for regular analysis pipelines adopting RAP principles we'd usually advise against using notebooks and instead recommend a Git controlled repository of code scripts using the [standard repository template available in dfeR](https://dfe-analytical-services.github.io/dfeR/reference/create_project.html). In addition to potential problems raised due to unfamiliarity with notebooks themselves, version control can also be more difficult in Databricks notebooks, as there is the potential for multiple people to make edits to the same notebook at once. Additionally, with a full script in an IDE, when the code is run without modifications it will provide the same output consistently. However, with notebooks you can run code cells independently of one another. If you don't run all code cells in order then your output could be inconsistent and result in errors, potentially making QA more difficult. This introduces more manual steps and increases the risk of human error. 
```

Comment:
I'd remove
>In addition to potential problems raised due to unfamiliarity with notebooks themselves,

It feels like it's labouring the point a little

Reply:
It felt weird not to acknowledge the callout thing above and then come up with a load of new, separate problems - but I agree

Reply:
Maybe something like this instead?

> For regular analysis pipelines and models adopting RAP principles we’d usually advise against using notebooks and instead recommend a Git controlled repository of code scripts using the [standard repository template available in dfeR](https://dfe-analytical-services.github.io/dfeR/reference/create_project.html). 
>
> We recommend this because in addition to notebooks being different to how many teams write code now, version control, collaboration and reproducibility can all be more difficult in Databricks notebooks.
> 
> In notebooks there is the potential for multiple people to make edits to the same notebook at once, making it much harder to version control and have a clear audit trail of what line has been changed and why.
>
> If you have multiple people working on a notebook, it is possible for your code run to be accidentally cancelled if another user starts running code in the same notebook.
>
> When you have a full script in an IDE, the script will provide the same output consistently. However, with notebooks you can run code cells independently of one another. If you don’t run all code cells in order then your output could be inconsistent and result in errors, potentially making QA more difficult. This introduces more manual steps and increases the risk of human error.